### PR TITLE
Fix missing auth headers with PATCH HTTP request when pushing to default port

### PR DESCRIPTION
**Pull request by stanhu**
If a user specifies `mydomain.com:443` in the `Host` configuration, the PATCH request for the layer upload will fail because the credentials will be omitted because the challenge does not appear to be in the map. To fix this, we normalize the map keys always to use the Host:Port combination.

Closes https://github.com/docker/docker/issues/18469

### CODE REVIEW

**Comment by RichardScothern** [comment]
can you rebase please @stanhu

**Comment by stanhu** [comment]
@RichardScothern Done.

**Comment by stanhu** [comment]
Ping @RichardScothern

File: registry/client/auth/addr.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+// FROM: http://golang.org/src/net/http/transport.go
+var portMap = map[string]string{
+	"http":  "80",
+	"https": "443",
+}
+
+// CanonicalAddr returns url.Host but always with a ":port" suffix
+// FROM: http://golang.org/src/net/http/transport.go
+func CanonicalAddr(url *url.URL) string {
```
Comment (RichardScothern) [inline registry/client/auth/addr.go]: This doesn't need to be exported unless you see it being called from another package

File: registry/client/auth/addr.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package auth
+
+import (
+	"net/url"
+	"strings"
+)
+
+// FROM: http://golang.org/src/net/http/client.go
```
Comment (dmcgowan) [inline registry/client/auth/addr.go]: Might have moved, found at https://golang.org/src/net/http/http.go

File: registry/client/auth/authchallenge.go [hunk 3]
```diff
@@ -70,11 +75,12 @@ func (m *simpleChallengeManager) AddResponse(resp *http.Response) error {
 	}
 	urlCopy := url.URL{
 		Path:   resp.Request.URL.Path,
-		Host:   strings.ToLower(resp.Request.URL.Host),
+		Host:   resp.Request.URL.Host,
 		Scheme: resp.Request.URL.Scheme,
 	}
 	m.Lock()
 	defer m.Unlock()
+	normalizeURL(&urlCopy)
```
Comment (dmcgowan) [inline registry/client/auth/authchallenge.go]: This doesn't need to be done in the lock

**Comment by dmcgowan** [comment]
@stanhu I would like to better understand the underlying issue that was run into. Was there only an error occurring for PATCH, and where was the `:443` port given by the user. Not sure which `Host` configuration you are referring to. If you have an HTTP trace that would be helpful, I am wondering whether there was a redirect that occurred during the initial ping that caused the URL given to `AddResponse` to differ from that used for `GetChallenges`.

**Comment by stanhu** [comment]
@dmcgowan See https://github.com/docker/docker/issues/18469 and https://github.com/docker/docker/issues/16728 for explanations of the bug.

Essentially, if you have a reverse proxy in front of the registry passing the `Host: xxx:443` headers, users will not be able to authenticate properly.

**Comment by dmcgowan** [comment]
I did read the issue but that is not helping me understand what is going on. There are certain parts of the issue which remain unclear and I want to fully understand so I can reproduce and test your fix. Your fix appears to the address the case where the response url added to challenge manager does not match the url being used to fetch the challenge. Since you have reproduced this error and assuming this change fixed your issue these questions may help me better understand what is going on...

1) Why is the reverse proxy changing the inbound host name to add the `:443` port?
2) Why does this only have an effect on PATCH/PUT requests?
3) Is there a redirect occurring during the initial ping to /v2/?
4) Does it only occur when the registry is also configured with the `:443` port?

**Comment by stanhu** [comment]
@dmcgowan Since this PR was created more than 2 months ago, my memory is a little hazy on the details, but I'll try my best to answer.

1) Why is the reverse proxy changing the inbound host name to add the :443 port?

In our case, we had an nginx template that was adding `X-Forwarded-For` for `Host:Port` to the nginx configuration file. If we included 443, then this would cause authentication to Docker Registry to fail in a non-obvious way as described in https://github.com/docker/docker/issues/18469. When we omitted 443, things worked fine. Note that if we moved the port to a non-standard port (e.g. 6000), things would work too. 

We worked around the problem by omitting standard ports, but this shouldn't be the way to do things.

2) Why does this only have an effect on PATCH/PUT requests?

I think it's because there is a redirect occurring if a layer does not exist causing a PATCH/PUT request to omit the required authorization header. See below.

3) Is there a redirect occurring during the initial ping to /v2/?

I don't have the HTTP logs on the server side, but from what I was seeing from the HTTPS logs in the client side:

In looking at a failed trace, this is what I see:

1. POST https://registry.com/v2/root/myrepository/blobs/uploads/ 

    This then returns a 202 with `Location: https://registry.com:443/v2/root/myrepository/blobs/uploads/<UUID>`, which indicates the layer was created.

2. PATCH `https://registry.com/v2/root/myrepository/blobs/uploads/<UUID>` <--- but this has no Authorization header!

4) Does it only occur when the registry is also configured with the :443 port?

Yes, as far as I know.

**Comment by stanhu** [comment]
CircleCI tests appear to be failing due to some networking connectivity issue.

File: registry/client/auth/addr.go (hunk 2 above)
Comment (stanhu) [inline registry/client/auth/addr.go]: Updated

File: registry/client/auth/authchallenge.go (hunk 3 above)
Comment (stanhu) [inline registry/client/auth/authchallenge.go]: Updated

**Comment by dmcgowan** [comment]
> PATCH https://registry.com/v2/root/myrepository/blobs/uploads/<UUID> <--- but this has no Authorization header!

Do you mean `PATCH https://registry.com:443/v2/root/myrepository/blobs/uploads/<UUID>` here? As I understand this issue now is that the Location returned by the registry has a host that differs from the original requested host. In that case your fix does make sense to address it. I would still like to highlight that this can be fixed with configuration on the other open issues and I don't believe the reverse proxy is relevant for this issue.

**Comment by stanhu** [comment]
> Do you mean PATCH https://registry.com:443/v2/root/myrepository/blobs/uploads/<UUID> here? 

`PATCH https://registry.com:443/v2/root/myrepository/blobs/uploads/<UUID>` is equivalent to `PATCH https://registry.com/v2/root/myrepository/blobs/uploads/<UUID>`, but the auth challenge code thinks they are different. As a result, when it goes to lookup the available authentication challenges, it does not find any matching authentication data for this URL, so it does not send the `Authorization` header.

> As I understand this issue now is that the Location returned by the registry has a host that differs from the original requested host. In that case your fix does make sense to address it. I would still like to highlight that this can be fixed with configuration on the other open issues and I don't believe the reverse proxy is relevant for this issue.

If the `Location` returned is truly different, I understand that this PR doesn't solve that problem. I welcome other ways to fix the problem, but since we've stopped including standard ports in headers, this bug is no longer an issue for us.

Yes, the reverse proxy is just an easy way to reproduce the issue.

File: registry/client/auth/addr.go (hunk 1 above)
Comment (stanhu) [inline registry/client/auth/addr.go]: Updated

**Comment by sudhanshu-garg** [comment]
I am still hitting the same issue even after picking up the above fix. See my comments here - https://github.com/docker/docker/issues/18469.

**Comment by sudhanshu-garg** [comment]
Copy paste from above:
----
In looking at a failed trace, this is what I see:

    POST https://registry.com/v2/root/myrepository/blobs/uploads/

    This then returns a 202 with Location: https://registry.com:443/v2/root/myrepository/blobs/uploads/<UUID>, which indicates the layer was created.

    PATCH https://registry.com/v2/root/myrepository/blobs/uploads/<UUID> <--- but this has no Authorization header!

    Does it only occur when the registry is also configured with the :443 port?
---

I think the fix should be to remove the standard port of 443 from the Location URL. In the above example it should have been and then it works perfectly fine:

    This then returns a 202 with Location: https://registry.com/v2/root/myrepository/blobs/uploads/<UUID>, which indicates the layer was created.

**Comment by stanhu** [comment]
I think https://registry.com and https://registry.com:443 should be equivalent, so I don't think removing the canonical port is the right fix.

I know this patch resolved the issue for me, but has it shipped with a Docker release yet? https://github.com/docker/distribution/commit/8cb75f692f1205f70b2fd5005bcde844efd8a41e says it's tagged in an RC for Docker Distribution 2.6.0.

**Comment by sudhanshu-garg** [comment]
>> I know this patch resolved the issue for me, but has it shipped with a Docker release yet?
Isn't the fix part of the docker-distribution code? If yes, I didn't quite understand your comment about has it shipped with a Docker release yet? I assume you are referring to Docker Distribution release. I cloned the latest master git repo and so I do have the fix locally but I still run into the exact same issue.

When you say the patch resolved the issue for you, was the location header getting returned still container the 443 port in it and even then the docker client/engine sent the Authorization header in the PATCH request?

**Comment by thaJeztah** [comment]
@sudhanshu-garg this patch contains code in the registry _client_, which is used by docker to communicate with the registry. Even though 443 is the standard port for https, IIRC, the docker client stores credentials by `registry[:port]`; without "protocol", so without this change, `my-registry` could either refer to `my-registry:443` or `my-registry:80`.

**Comment by sudhanshu-garg** [comment]
Ah, thanks for the clarification @thaJeztah. I get it now. So the patch fixes the registry client. I didn't know the docker-distribution project contains Go code which gets bundled with docker client/engine released as part of the Docker releases.

**Pull request merged**